### PR TITLE
Fixes #13959 - Add support to boot from volume snapshot in openstack

### PR DESCRIPTION
**Pull request by ShimiTaNaka**

### CODE REVIEW

**Comment by lzap** [comment]
Hello, thank you for your contribution. I will take a look on your patch and test it against my instance.

File: app/views/compute_resources_vms/form/openstack/_base.html.erb [hunk 1]
```diff
@@ -3,13 +3,21 @@
...
 <div id='image_selection'>
   <%= select_f f, :image_ref, images, :uuid, :name,
                { :include_blank => (images.empty? || images.size == 1) ? false : _('Please select an image') },
                { :disabled => images.empty?, :label => _("Image"), :label_size => "col-md-2" } %>
 </div>
+<div id='snapshot_selection'>
+  <%= checkbox_f f, :clone_from_snapshot, {:label => "Clone from snapshot", :help_inline => _("Use snapshot instead of image"), :id => 'use_snapshot_for_vm'}, "true", "false" %>
+  <%= select_f f, :snapshot_id, snapshots, :id, :name,
+               { :include_blank => (snapshots.empty? || snapshots.size == 1) ? false : _('Please select a snapshot') },
+               { :disabled => snapshots.empty?, :label => _("Snapsot") } %>
```
Comment (lzap) [inline app/views/compute_resources_vms/form/openstack/_base.html.erb]: Typo.

I tried your patch on Foreman develop and RHOS Kilo and my list of snapshots does not populate at all. Is there a patch for fog I need to have?

Comment (ShimiTaNaka) [inline app/views/compute_resources_vms/form/openstack/_base.html.erb]: Issue re-produced with creating a snapshot from CLI, when creating with CLI and no params, the name is set to None and the fog returns nil in the name attribute.
Snapshot needs a name to be displayed, can default it to the ID if name is not present if we want.

**Comment by ShimiTaNaka** [comment]
I've added a fix to show the ID instead of the name if name is not available (The way the UI of openstack shows it)

**Comment by tbrisker** [comment]
[test]

**Comment by lzap** [comment]
Nope, does not show me snapshots.

```
[lzap@lzapx foreman]$ bi | grep fog
Using fog-core 1.36.0
Using fog-json 1.0.2
Using fog-aliyun 0.1.0
Using fog-xml 0.1.2
Using fog-atmos 0.1.0
Using fog-aws 0.9.2
Using fog-brightbox 0.10.1
Using fog-cloudatcost 0.1.2
Using fog-dynect 0.0.3
Using fog-ecloud 0.3.0
Using fog-google 0.1.0
Using fog-local 0.3.0
Using fog-openstack 0.1.2
Using fog-powerdns 0.1.1
Using fog-profitbricks 0.0.5
Using fog-rackspace 0.1.1
Using fog-radosgw 0.0.5
Using fog-riakcs 0.1.0
Using fog-sakuracloud 1.7.5
Using fog-serverlove 0.1.2
Using fog-softlayer 1.1.0
Using fog-storm_on_demand 0.1.1
Using fog-terremark 0.1.0
Using fog-vmfusion 0.1.0
Using fog-voxel 0.1.0
Using fog-vsphere 0.6.3
Using fog-xenserver 0.2.3
Using fog 1.38.0
Using fog-libvirt 0.1.0
```

Let's start by comparing our versions?

**Comment by ShimiTaNaka** [comment]
$ bundle install --without=postgresql | grep fog
Using fog-core 1.36.0
Using fog-json 1.0.2
Using fog-xml 0.1.2
Using fog-local 0.3.0
Using fog-vmfusion 0.1.0
Using fog-vsphere 0.6.3
Using fog-aliyun 0.1.0
Using fog-brightbox 0.10.1
Using fog-sakuracloud 1.7.5
Using fog-serverlove 0.1.2
Using fog-softlayer 1.1.0
Using fog-storm_on_demand 0.1.1
Using fog-atmos 0.1.0
Using fog-aws 0.9.2
Using fog-cloudatcost 0.1.2
Using fog-dynect 0.0.3
Using fog-ecloud 0.3.0
Using fog-google 0.1.0
Using fog-openstack 0.1.2
Using fog-powerdns 0.1.1
Using fog-profitbricks 0.0.5
Using fog-rackspace 0.1.1
Using fog-radosgw 0.0.5
Using fog-riakcs 0.1.0
Using fog-terremark 0.1.0
Using fog-voxel 0.1.0
Using fog-xenserver 0.2.3
Using fog-libvirt 0.1.0
Using fog 1.38.0

Identical.

**Comment by lzap** [comment]
This does not seem to work with our Red Hat OpenStack instance apparently, maybe some extra security settings. I can only see empty list coming out from fog. Maybe someone else with own OpenStack instance can try this patch? I know @dLobatog and @GregSutcliffe have one? Folks?

Since you added snapshot support, would you mind creating new tab in the Compute Resource page along with Images so user can easily list all Snapshots available there? It should be relatively easy since you have the model method now.

**Comment by ShimiTaNaka** [comment]
The Images tab is to create an instance of the image model [Image](https://github.com/myheritage/foreman/blob/develop/app/models/image.rb)
didn't quite understand why need to create an image BOTH in openstack AND foreman but ok...
If we apply the same logic, it will require adding a new model as well I assume...

**Comment by GregSutcliffe** [comment]
Sorry @lzap I recycled my Openstack devbox ages ago - I'd be testing against OS1, just like you.

**Comment by lzap** [comment]
Okay I was able to setup a free account on trystack.org and it looks like it works just fine there. I am able to review and test now.

First of all, I dont like another checkbox, because this is not consistent. We should follow UI design of the Launch instance of the OpenStack UI which is a drop down with the following items:
- boot  from image
- boot  from snapshot
- boot  from volume
- boot  from image (create new volume)
- boot  from volume snapshot (create new volume)

I think we should do the same, either dropdown or radio group would be appropriate. Also the snapshot dropdown should be disabled, until the selection is made to boot from snapshot. This is my first remark from the UI perspective.

File: app/models/compute_resources/foreman/model/openstack.rb [hunk 2]
```diff
@@ -94,7 +107,7 @@ def boot_from_volume(args = {})
     end
 
     def create_vm(args = {})
-      boot_from_volume(args) if Foreman::Cast.to_bool(args[:boot_from_volume])
+      boot_from_volume(args) if (Foreman::Cast.to_bool(args[:boot_from_volume]) || Foreman::Cast.to_bool(args[:boot_from_snapshot]))
```
Comment (lzap) [inline app/models/compute_resources/foreman/model/openstack.rb]: Don't abuse the helper mehod, create your own one `boot_from_snapshot` I guess.

Comment (lzap) [inline app/models/compute_resources/foreman/model/openstack.rb]: Or reuse it and rename it to something more appropriate like `create_boot_volume` perhaps?

**Comment by lzap** [comment]
I think you should do the UI as either dropdown or radiobox with the following options:
- boot from image
- boot from image (create new volume)
- boot from volume snapshot (create new volume)

The first one should be the default.

**Pull request closed without merging**